sdk/php/src/Command/EntrypointCommand.php

```diff
@@ -0,0 +1,239 @@
+class EntrypointCommand extends Command
+{
+    private Dagger\Client $daggerClient;
+
+    public function __construct()
+    {
+        parent::__construct();
+        $this->daggerClient = Dagger\Dagger::connect();
+    }
+
+    protected function execute(
+        InputInterface $input,
+        OutputInterface $output
+    ): int {
+        $functionCall = $this->daggerClient->currentFunctionCall();
+        $parentName = $functionCall->parentName();
+
+        try {
+                return $parentName === '' ?
+                    $this->registerModule($functionCall) :
+                    $this->callFunctionOnParent(
+                        $output,
+                        $functionCall,
+                        $parentName,
+                    );
+        } catch (\Throwable $t) {
+            $this->outputErrorInformation($input, $output, $t);
+
+            return Command::FAILURE;
+        }
+    }
+
+    private function registerModule(Dagger\FunctionCall $functionCall): int
+    {
+        $daggerModule = $this->daggerClient->module();
+
+        $src = (new FindsSrcDirectory())();
+        $daggerObjects = (new FindsDaggerObjects())($src);
+
+        foreach ($daggerObjects as $daggerObject) {
+            $objectTypeDef = $this->daggerClient
+                ->typeDef()
+                ->withObject($this->normalizeClassname($daggerObject->name));
+
+            foreach ($daggerObject->daggerFunctions as $daggerFunction) {
+                $func = $this->daggerClient->function(
+                    $daggerFunction->name,
+                    $this->getTypeDef($daggerFunction->returnType)
+                );
+
+                if ($daggerFunction->description !== null) {
+                    $func = $func->withDescription($daggerFunction->description);
+                }
+
+                foreach ($daggerFunction->arguments as $argument) {
+                    $func = $func->withArg(
+                        $argument->name,
+                        $this->getTypeDef($argument->type),
+                        $argument->description,
+                        $argument->default
+                    );
+                }
+
+                $objectTypeDef = $objectTypeDef->withFunction($func);
+            }
+
+            $daggerModule = $daggerModule->withObject($objectTypeDef);
+        }
+
+        $functionCall->returnValue(new Dagger\Json(json_encode(
+            (string) $daggerModule->id()
+        )));
+
+        return Command::SUCCESS;
+    }
+
+    private function callFunctionOnParent(
+        OutputInterface $output,
+        Dagger\FunctionCall $functionCall,
+        string $parentName
+    ): int {
+        $errorOutput = $output instanceof ConsoleOutputInterface ?
+            $output->getErrorOutput() :
+            $output;
+
+        $className = "DaggerModule\\$parentName";
+        $functionName = $functionCall->name();
+        $class = new $className();
+        $class->client = $this->daggerClient;
```

Comment:
Just a note for later.

By convention, our SDKs provide a **global** client connection under the name `dag`. See https://dagger.io/blog/simpler-reusability-with-the-global-client.

## Python

https://github.com/dagger/dagger/blob/aef8596314ad777085f11580c6911346f57c806c/sdk/python/src/dagger/client/gen.py#L7100-L7101

## TypeScript

https://github.com/dagger/dagger/blob/aef8596314ad777085f11580c6911346f57c806c/sdk/typescript/api/client.gen.ts#L9205

## Go

https://github.com/dagger/dagger/blob/aef8596314ad777085f11580c6911346f57c806c/cmd/codegen/generator/go/templates/src/_dagger.gen.go/defs.go.tmpl#L143-L151

Reply:
For consistency I've now renamed the `daggerClient` variable in the `EntryPointCommand` to `dag`.

The current solution of `$class->client = $this->daggerClient;` is more of a quick hack to get it working than a final implementation. 

Global variables don't tend to be used often in PHP so may be confusing for PHP developers. The plan is to have the SDK automatically inject the `Client` dependency into `DaggerObjects` requiring it. We plan to implement this shortly after sorting out custom objects.

Reply:
We have now also defined `DAG` as a constant, making it globally accessible.

It has been named in uppercase as that is the standard PHP naming convention for constants.

https://github.com/dagger/dagger/blob/cb6b89da641d2d883682ba037951f32ea3b3f678/sdk/php/template/dagger#L20-L27

Reply:
> Global variables don't tend to be used often in PHP so may be confusing for PHP developers.

I know I said "global" but it doesn't have to be a literal `global` variable. It can (and should) be imported. It just needs to setup it's own connection automatically and return a (shared) client instance.

Reply:
We now have `sdk/php/src/functions.php` which lets us import the client like so:

```php
//...

use function Dagger\dag;

#[DaggerObject]
class Example
{
     #[DaggerFunction('Echo the value to standard output')]
     public function echo(string $value): Container
     {
         return dag()
             ->container()
             ->from('alpine:latest')
             ->withExec(['echo', $value]);
     }

// ...
```